src/Google.Api.Gax/Api/Gax/GetOrCreateHelper.cs

```diff
@@ -0,0 +1,153 @@
+        /// (whole get/create cycles) is limited by <paramref name="maxAttempts"/>.
+        /// </para>
+        /// <para>
+        /// If <paramref name="validator"/> is not <c>null</c>, it is called when a resource is fetched rather than created,
+        /// with the fetched resource being passed in. If the result of the validation is non-null, it is used as the message
+        /// of a <see cref="ResourceMismatchException"/> which is thrown as the result of the operation.
+        /// </para>
+        /// </remarks>
+        /// <typeparam name="TResource">The type of the resource being fetched or created.</typeparam>
+        /// <param name="fetcher">A delegate to fetch the resource. Must not be null.</param>
+        /// <param name="creator">A delegate to create the resource if it doesn't already exist. Must not be null.</param>
+        /// <param name="validator">An optional delegate to validate that a fetched resource has the expected form.</param>
+        /// <param name="maxAttempts">The maximum number of fetch/create cycles to try before giving up.</param>
+        /// <exception cref="ResourceMismatchException">The resource was fetched, but failed the specified validation.</exception>
+        /// <returns>The resource fetched or created.</returns>
+        public static TResource GetOrCreate<TResource>(
+            Func<TResource> fetcher,
+            Func<TResource> creator,
+            Func<TResource, string> validator = null,
+            int maxAttempts = DefaultMaxAttempts)
+        {
+            GaxPreconditions.CheckNotNull(fetcher, nameof(fetcher));
+            GaxPreconditions.CheckNotNull(creator, nameof(creator));
+            GaxPreconditions.CheckArgumentRange(maxAttempts, nameof(maxAttempts), 1, int.MaxValue);
+            for (int i = 0; i < maxAttempts; i++)
+            {
+                try
+                {
+                    var resource = fetcher();
+                    var validationError = validator?.Invoke(resource);
+                    if (validationError != null)
+                    {
+                        throw new ResourceMismatchException(validationError);
+                    }
+                    return resource;
+                }
+                catch (RpcException e) when (e.Status.StatusCode == StatusCode.NotFound)
+                {
+                    try
+                    {
+                        return creator();
+                    }
+                    catch (RpcException creationException) when (i < maxAttempts - 1 && creationException.Status.StatusCode == StatusCode.AlreadyExists)
+                    {
+                        // Probably a race condition... someone else tried to create it just when we did.
+                        // Go back through the loop and check we can get it this time, failing after this happens too often.
+                    }
```

Comment:
Should we have a bit of a wait before going back to see if we can fetch the result? Wondering there can be a bit of a delay between the creation of a resource and been able to fetch it.